sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -1388,4 +1410,208 @@ public void deidentify(ProcessContext context)
+     * will output a {@link PCollectionTuple} which contains the output and dead-letter {@link
+     * PCollection}*.
+     *
+     * <p>The {@link PCollectionTuple} output will contain the following {@link PCollection}:
+     *
+     * <ul>
+     *   <li>{@link FhirIO.Search#OUT} - Contains all {@link PCollection} records successfully search
+     *       from the Fhir store.
+     *   <li>{@link FhirIO.Search#DEAD_LETTER} - Contains all {@link PCollection} of {@link
+     *       HealthcareIOError}* of failed searches from the Fhir store, with
+     *       error message and stacktrace.
+     * </ul>
+     */
+    static class SearchResourcesJsonString
+            extends PTransform<PCollection<KV<String, Map<String, String>>>, FhirIO.Search.Result> {
+
+      private final ValueProvider<String> fhirStore;
+
+      /** Instantiates a new Search Fhir resources DoFn. */
+      public SearchResourcesJsonString(ValueProvider<String> fhirStore) {
+        this.fhirStore = fhirStore;
+      }
+
+      @Override
+      public FhirIO.Search.Result expand(PCollection<KV<String, Map<String, String>>> resourceIds) {
+        return new FhirIO.Search.Result(
+                resourceIds.apply(
+                        ParDo.of(new SearchResourcesFn(this.fhirStore))
+                                .withOutputTags(FhirIO.Search.OUT, TupleTagList.of(FhirIO.Search.DEAD_LETTER))));
+      }
+
+      /** DoFn for searching messages from the Fhir store with error handling. */
+      static class SearchResourcesFn extends DoFn<KV<String, Map<String, String>>, String> {
+
+        private Counter failedSearches =
+                Metrics.counter(SearchResourcesFn.class, "failed-fhir-searches");
+        private static final Logger LOG = LoggerFactory.getLogger(SearchResourcesFn.class);
+        private final Counter successfulSearches =
+                Metrics.counter(SearchResourcesFn.class, "successful-fhir-searches");
+        private HealthcareApiClient client;
+        private final ValueProvider<String> fhirStore;
+
+        /** Instantiates a new Fhir resources search fn. */
+        SearchResourcesFn(ValueProvider<String> fhirStore) {
+          this.fhirStore = fhirStore;
+        }
+
+        /**
+         * Instantiate healthcare client.
+         *
+         * @throws IOException the io exception
+         */
+        @Setup
+        public void instantiateHealthcareClient() throws IOException {
+          this.client = new HttpHealthcareApiClient();
+        }
+
+        /**
+         * Process element.
+         *
+         * @param context the context
+         */
+        @ProcessElement
+        public void processElement(ProcessContext context) {
+          KV<String, Map<String, String>> elementValues = context.element();
+          try {
+            context.output(searchResources(
+                    this.client, this.fhirStore.toString(), elementValues.getKey(), elementValues.getValue()));
+          } catch (Exception e) {
```

Comment:
Please only catch the exceptions you expect to handle.

Reply:
Would that mean that unhandled exceptions / errors won't be logged?

Reply:
They will be handled by code that knows how to handle them. For example, the code here probably should not be responsible for an OutOfMemoryError.

Reply:
Ah, makes sense, done. Sorry Java newbie here!